### PR TITLE
C API/Example: add lists c api, and enhance c api and example.

### PR DESCRIPTION
**Pull request by sherry-1001**
Signed-off-by: zhichen.jiang <zhichen.jiang@intel.com>

What's Changed:

   Add c api for lists data type. And enhance c api and example.

Tests 

- Unit test
- Integration test
    add example.
- No code

Side effects
     no

### CODE REVIEW

File: engine/c.cc [hunk 1]
```diff
@@ -197,6 +241,8 @@ KVDKIterator *KVDKCreateIterator(KVDKEngine *engine, const char *collection,
   } else if (iter_type == HASH) {
     result->rep =
         (engine->rep->NewUnorderedIterator(std::string(collection))).get();
+  } else {
+    return nullptr;
   }
   assert(result->rep != nullptr && "Create Iterator Failed!");
```
Comment (ZiyanShi) [inline engine/c.cc:247]: Should also return nullptr here?

Comment (ZiyanShi) [reply engine/c.cc:247]: It's not up to user to check result->rep. User only needs to check result != nullptr.

File: examples/tutorial/c_api_tutorial.c [hunk 2]
```diff
@@ -12,6 +12,286 @@
 // Modify this path if necessary.
 const char *pmem_path = "/mnt/pmem0/tutorial_kvdk_example";
 
+static int CmpCompare(const char *a, size_t alen, const char *b, size_t blen) {
```
Comment (ZiyanShi) [inline examples/tutorial/c_api_tutorial.c:15]: StrCmp or StringCompare

File: include/kvdk/engine.h [hunk 3]
```diff
@@ -34,6 +34,7 @@ typedef struct KVDKWriteBatch KVDKWriteBatch;
 typedef struct KVDKIterator KVDKIterator;
 
 typedef enum { SORTED, HASH } KVDKIterType;
+typedef enum {LEFT, RIGHT} KVDKListOpType
```
Comment (ZiyanShi) [inline include/kvdk/engine.h:37]: This thing should not be exposed to user. It's a internal thing.

File: include/kvdk/engine.h [hunk 4]
```diff
@@ -103,18 +109,40 @@ extern KVDK_LIBRARY_API KVDKStatus KVDKHashGet(KVDKEngine *engine,
                                                const char *key, size_t key_len,
                                                size_t *val_len, char **val);
 
+// For Queue
+extern KVDK_LIBRARY_API KVDKStatus KVDKLPush(KVDKEngine *engine,
+                                             const char *collection,
+                                             size_t collection_len,
+                                             const char *key, size_t key_len);
+extern KVDK_LIBRARY_API KVDKStatus KVDKLPop(KVDKEngine *engine,
```
Comment (ZiyanShi) [inline include/kvdk/engine.h:117]: Maybe we should use a String struct to pack these parameters.

**Review by ZiyanShi — CHANGES_REQUESTED** [review]

File: include/kvdk/engine.h (hunk 3 above)
Comment (sherry-1001) [reply include/kvdk/engine.h:37]: It 's my old code, I forgot deleting it. Thanks for reminder.

File: include/kvdk/engine.h (hunk 4 above)
Comment (sherry-1001) [reply include/kvdk/engine.h:117]: Aligned c++ API. Talked with @KVKIT , the current API maybe be more clear.

File: examples/tutorial/c_api_tutorial.c (hunk 2 above)
Comment (sherry-1001) [reply examples/tutorial/c_api_tutorial.c:15]: not limited string. Reference leveldb: https://github.com/google/leveldb/blob/master/db/c_test.c#L107

File: engine/c.cc (hunk 1 above)
Comment (sherry-1001) [reply engine/c.cc:247]: ok

**Review by ZiyanShi — APPROVED** [review]

**Pull request merged**